### PR TITLE
Fix sparse compressed tensor invariants checks when nnz==0

### LINKED ISSUE #115755: Sparse compressed tensor invariants checking fail when nnz == 0
## Issue description

As in the title.

## Code example 1 (false positive result)

```
>>> a = torch.sparse_csr_tensor(torch.tensor([123, 0]), torch.tensor([], dtype=torch.int64), [], (1, 1), check_invariants=True)
```
succeeds while `crow_indices` content is invalid. Expected a failure from sparse tensor invariant checks.

## Code example 2 (python crash)

```python
>>> a = torch.sparse_csr_tensor(torch.tensor([0, 0]), torch.tensor([], dtype=torch.int64), [], (1, 1), check_invariants=True)
>>> torch._validate_compressed_sparse_indices(True, a.crow_indices(), a.col_indices(), a.shape[0], a.shape[1], a._nnz())
Floating point exception (core dumped)
```
Expected no crash and success from the invariants check.

## System Info

- PyTorch version: main

### PR DESCRIPTION
**Pull request by pearu**
Fixes https://github.com/pytorch/pytorch/issues/115755

This PR is a step toward deprecating `torch.empty(..., layout=<sparse compressed tensor layout>)` that usage should be minimized as it will produce invalid tensors, see also https://github.com/pytorch/pytorch/issues/90695 .

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115826

cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @jcaip

### CODE REVIEW

File: aten/src/ATen/native/TensorFactories.cpp [hunk 1]
```diff
@@ -1277,14 +1278,51 @@ Tensor triu_indices_cpu(
...
+                                              pin_memory);
+}
+
 Tensor zeros_symint(SymIntArrayRef size,
     c10::optional<ScalarType> dtype,
     c10::optional<Layout> layout,
     c10::optional<Device> device,
     c10::optional<bool> pin_memory) {
+  Layout layout_ = layout.value_or(Layout::Strided);
+  if (at::sparse_csr::is_sparse_compressed(layout_)) {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/TensorFactories.cpp:1321]: I'm not sure we can land this due to overhead. Can you use the dispatcher for this to register a new function?

**Review by cpuhrsch — APPROVED** [review]
Great! Do we also check that values.numel() == 0 for the zeros constructor? I think it's implied by the tests, but just double checking to make sure.

File: aten/src/ATen/native/TensorFactories.cpp (hunk 1 above)
Comment (pearu) [reply aten/src/ATen/native/TensorFactories.cpp:1321]: No. `torch.zeros` is factory function that takes no tensor arguments, hence the dispatcher mechanism is not applicable.

@cpuhrsch what kind of overhead are you referring here? `is_sparse_compressed` is a rather simple inline function containing a simple switch statement.
One possible optimization would be:
```c++
if (layout_ != Layout::Strided && at::sparse_csr::is_sparse_compressed(layout_)) {
  ...
}
... // original code
```
but I am not sure that there will be a considerable overhead in either cases.

**Comment by pearu** [comment]
> Great! Do we also check that values.numel() == 0 for the zeros constructor? I think it's implied by the tests, but just double checking to make sure.

`values` shape is `plain_indices_and_values_size` that has the last dimension equal to `nnz` (`=0`). Since `numel` is computed as product of dimensions, the result of `values.numel()` is always `0` from the `zeros` constructor.

On the other hand, sparse tensor invariants checks (that are always executed when running tests) ensure that `values.shape[-1]` is equal to `crow_indices[-1]` and the `zeros` constructor calls `crow_indices.zero_()`. So, the answer is affirmative.

File: aten/src/ATen/native/TensorFactories.cpp (hunk 1 above)
Comment (amjames) [reply aten/src/ATen/native/TensorFactories.cpp:1321]: Yeah there should be no concerns with additional overhead from these change.

**Review by amjames — APPROVED** [review]

**Pull request closed without merging**